src/depth_limit.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017, 2018 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::codec::{Error, Decode, Input};
+
+/// The error message returned when depth limit is reached.
+const DECODE_MAX_DEPTH_MSG: &str = "Maximum recursion depth reached when decoding";
+
+/// Extension trait to [`Decode`] that ensures that the given input data is consumed completly while
+/// decoding.
+pub trait DecodeLimit: Sized {
+	/// Decode `Self` with given maximum recursion depth.
+	/// is returned.
+	fn decode_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error>;
+	/// Decode `Self` and consume all of the given input data. If not all data is consumed, an error
+	/// is returned.
+	fn decode_all_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error>;
+}
+
+
+struct DepthTrackingInput<'a, I> {
+	input: &'a mut I,
+	depth: u32,
+	max_depth: u32,
+}
+
+impl<'a, I:Input> Input for DepthTrackingInput<'a, I> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, Error> {
+		self.input.remaining_len()
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), Error> {
+		self.input.read(into)
+	}
+
+	fn read_byte(&mut self) -> Result<u8, Error> {
+		self.input.read_byte()
+	}
+
+	fn descend_ref(&mut self) -> Result<(), Error> {
+		self.input.descend_ref()?;
+		self.depth += 1;
+		if self.depth > self.max_depth {
+			Err(DECODE_MAX_DEPTH_MSG.into())
+		} else {
+			Ok(())
+		}
+	}
+
+	fn ascend_ref(&mut self) {
+		self.input.ascend_ref();
+		self.depth -= 1;
+	}
+}
+
+impl<T: Decode> DecodeLimit for T {
+	fn decode_all_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error> {
+		let mut input = DepthTrackingInput {
+			input: &mut &input[..],
+			depth: 0,
+			max_depth: limit,
+		};
+		let res = T::decode(&mut input)?;
+
+		if input.input.is_empty() {
+			Ok(res)
+		} else {
+			Err(crate::decode_all::DECODE_ALL_ERR_MSG.into())
+		}
+	}
+
+	fn decode_with_depth_limit(limit: u32, input: &[u8]) -> Result<Self, Error> {
+		let mut input = DepthTrackingInput {
+			input: &mut &input[..],
```

Comment:
This here creates a new byte slice that's separate from `input`, which would not advance the `input` index after decoding is done.

Reply:
That would require changing the signature to accept a mutable input. If such a use case is required a PR is welcome